drodrpg/DRODLib/DbSavedGames.cpp

```diff
@@ -3113,6 +3119,129 @@ void CDbSavedGames::UpdatePlayerTallies(
+				break;
+			}
+		}
+	}
+	delete pMapStateSave;
+	return mapStates;
+}
+
+//******************************************************************************
+void CDbSavedGames::UpdateTotalMapStates(const UINT dwPlayerID, const RoomMapStates& RoomMapStates)
+//Update ST_TotalMapStates
+{
+	// No backwards search is to ensure this works when merging a player import, but it'll likely be an early db entry anyway
+	const UINT dwID = FindByType(ST_TotalMapStates, dwPlayerID, false);
+	CDbSavedGame* pMapStateSave = GetByID(dwID);
+	if (!pMapStateSave)
+	{
+		//Make new save record to store player's room map states.
+		pMapStateSave = GetNew();
+		pMapStateSave->dwPlayerID = dwPlayerID;
+		pMapStateSave->eType = ST_TotalMapStates;
+		pMapStateSave->bIsHidden = true;
+	}
+
+	//Merge RoomMapStates list with existing list.
+	bool bUpdate = false;
+	for (RoomMapStates::const_iterator it = RoomMapStates.begin(); it != RoomMapStates.end(); ++it)
+	{
+		const UINT roomID = it->first;
+		ASSERT(roomID);
+		vector<ExploredRoom*>::const_iterator room;
+		for (room = pMapStateSave->ExploredRooms.begin(); room != pMapStateSave->ExploredRooms.end(); ++room)
+		{
+			ExploredRoom* pRoom = *room;
+			if (pRoom->roomID == roomID)
+			{
+				if (pRoom->mapState < it->second)
+				{
+					pRoom->mapState = it->second;
+					bUpdate = true;
+				}
+				break;
+			}
+		}
+
+		//Room not found -- add a record of it.
+		if (room == pMapStateSave->ExploredRooms.end())
+		{
+			ExploredRoom* pExpRoom = new ExploredRoom();
+			pExpRoom->roomID = roomID;
+			pExpRoom->mapState = it->second;
+			pMapStateSave->ExploredRooms.push_back(pExpRoom);
+			bUpdate = true;
+		}
+	}
+
+	if (bUpdate)
+		pMapStateSave->Update();
+
+	delete pMapStateSave;
+}
+
+//*******************************************************************************
+void CDbSavedGames::UpdateTotalMapStatesWithAllSavedGameRooms(const UINT dwPlayerID)
+//Update ST_TotalMapStates with explored rooms from every save
+{
+	RoomMapStates mapStates;
+	CDb db;
+	db.SavedGames.FilterByPlayer(dwPlayerID);
+	db.SavedGames.FindHiddens(true);
+	const CIDSet savedGameIDs = db.SavedGames.GetIDs();
+
+	for (CIDSet::const_iterator iter = savedGameIDs.begin(); iter != savedGameIDs.end(); ++iter)
+	{
+		CDbSavedGame* pSavedGame = db.SavedGames.GetByID(*iter);
+		ASSERT(pSavedGame);
+		if (pSavedGame->eType != ST_DemoUpload && //Not a real saved game record
+			pSavedGame->eType != ST_TotalMapStates) //We will merge with this later
+		{
+			for (vector<ExploredRoom*>::const_iterator room = pSavedGame->ExploredRooms.begin(); room != pSavedGame->ExploredRooms.end(); ++room)
+			{
+				ExploredRoom* pRoom = *room;
+				MapState currentMapState = MapState::Invisible;
+				RoomMapStates::const_iterator currentMapStateIter = mapStates.find(pRoom->roomID);
+				if (currentMapStateIter != mapStates.end()) {
+					currentMapState = currentMapStateIter->second;
+				}
+				if (pRoom->mapState > currentMapState)
```

Comment:
The ">" operator check feels potentially fragile for long-term maintenance if new values ever get added to MapState.

I'm wondering whether it makes sense to refactor comparisons like this, which implicitly assume higher values are the ones we want to retain, into a method called something like "IsMoreDetailedRoomState" that encapsulates this logic. Reason being is that it may not be true at some point in the future that a greater value always indicates an "improved" room state that we want to track here.

WDYT?

Reply:
Sounds like a good idea to me. I've made this change.